### PR TITLE
[SPARK-28853][SQL] Support conf to organize file partitions by file path

### PR DESCRIPTION
**Pull request by triplesheep**
### What changes were proposed in this pull request?
This patch upports a conf to organize filePartition by file path. 

### Why are the changes needed?
When dynamicly writing data to hdfs it may generates a lot of small files, so sometimes we need to merge those files. When reading this files and writing again, it will be helpful if the read file RDD partitions is formed by partitions on hdfs.

Currently in FileSourceScanExec.createNonBucketedReadRDD after spliting files, spark will sort files with file size so it may scatter the partition distribution of the data files. It is a great help to support sort by file path here.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?

New test in FileSourceStrategySuite.scala is added.

### CODE REVIEW

**Comment by triplesheep** [comment]
cc @srowen @dongjoon-hyun @gengliangwang

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -480,7 +480,16 @@ case class FileSourceScanExec(
           partitionValues = partition.values
         )
       }
-    }.sortBy(_.length)(implicitly[Ordering[Long]].reverse)
+    }
+
+    val orderByFilePath = fsRelation.sparkSession.sessionState.conf.orderByFilePathWhenPartitioning
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: `conf.orderByFilePathWhenPartitioning`?

**Comment by maropu** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 1 above)
Comment (triplesheep) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: @maropu Do you mean this name is not suitable?

**Review by srowen — COMMENTED** [review]
Hm, I'm uneasy that this may introduce behavior and performance changes, but I don't have a specific problem already in mind. Why would name necessarily be more meaningful?

**Comment by dongjoon-hyun** [comment]
Hi, @triplesheep . Thank you for contribution and pining me.

I understand the intention, but I'm negative for this approach because this is too specific due to two hidden assumptions; (1) the original file contents are ordered by the file names and (2) the file name ordering will generate the best compression ratio.

> When dynamicly writing data to hdfs it may generates a lot of small files, so sometimes we need to merge those files. When reading this files and writing again, it will be helpful if the read file RDD partitions is formed by partitions on hdfs.

However, the files are not a single column data in many cases. In most cases, you will have more complex schema ORC/Parquet files. Given that, the original writing order doesn't guarantee the best merging way. There are various optimization possibilities if you choose the proper sorting columns from those small files during rewriting.

In short, the current approach seems to focus only simply merging files, but that looks not worth to have a new conf because it's not a best way to reorganize the small files.

cc @gatorsmile , @cloud-fan

**Comment by dongjoon-hyun** [comment]
A few second ago, I left a comment but deleted it. Sorry about that.

**Comment by gengliangwang** [comment]
@triplesheep Thanks for the work. Could you provide benchmark result to prove that it is helpful in certain scenarios?

**Comment by triplesheep** [comment]
> Hm, I'm uneasy that this may introduce behavior and performance changes, but I don't have a specific problem already in mind. Why would name necessarily be more meaningful?

Yeah, it exactly may introduce behavior and performance changes so I set it to be a SQLConf with default value false. The file path is meaning here because it represents the partitions of data, for example /base_path/day=1/file1, /base_path/day=2/file2 from the path we can know file1 and file2 are located in different partitions.

**Comment by triplesheep** [comment]
> A few second ago, I left a comment but deleted it. Sorry about that.

Never mind :)

**Comment by triplesheep** [comment]
> @triplesheep Thanks for the work. Could you provide benchmark result to prove that it is helpful in certain scenarios?

I have add a test for the change, it can prove it works lining with expectation.

**Comment by gengliangwang** [comment]
> I have add a test for the change, it can prove it works lining with expectation.
I meant a real workload that shows the total time of file scan is reduced with the new conf enabled.

**Comment by cloud-fan** [comment]
I think it's fragile to assume that similarly-named paths tend to store in the same location.

**Comment by triplesheep** [comment]
@srowen  @cloud-fan  emmm, actually I don't assume that similarly-named paths tend to store in the same location. Here is the scene it for small file merge.
1. Think about partition table: CREATE TABLE table_name PARTITIONED BY (day)
2. When we write data into that table we use dynamic write interface and it will write out more than one files in one spark task. So it may results on a lot of small files.
3. We want to merge those small files, so we read those data again and want to write it down to the table with less files.
4. When we read those small files back again, we want those files organized by partition(in my case is day) in FileRDD, so when write it back to the table, each RDD partition will write by a task and it will only write out in one files by dynamic write interface. So finally we merge the small files.

We can do this by read files again and use repartition but it was expensive so I post this issue.
I don't concern about the similarly-named paths store where I focus on the similarly-named paths store on the same RDD partition when read data from files :)

**Comment by triplesheep** [comment]
> > I have add a test for the change, it can prove it works lining with expectation.
> 
> I meant a real workload that shows the total time of file scan is reduced with the new conf on comparing when it is off.

Get that :)

**Comment by cloud-fan** [comment]
> When we read those small files back again, we want those files organized by partition(in my case is day) in FileRDD, so when write it back to the table, each RDD partition will write by a task and it will only write out in one files by dynamic write interface. So finally we merge the small files.

Can't we do a `partitionBy` before writing a DataFrame?

**Comment by triplesheep** [comment]
Hi, @cloud-fan :) we can use partitionBy to do this but I don't do that for two reasons:
1. partitionBy will be expensive because of shuffle
2. I don't want all data in a partition ends in the same RDD partition because it may be very large and the write may be slow. It seems like data skew. When use partitionBy it will put same key into same RDD partition and will encounter the problem I said above.

I want one RDD partition only contains the data from one data partition doesn't means I want data in same data partitions locates into the same RDD partition. For example:
Data Distribution:
day = 1: file1  file2 file3 file4
day = 2: file5
day = 3: file6
What I want in RDD partition(split by conf FILES_MAX_PARTITION_BYTES):
RDD partition1: file1 file2   (day=1)
RDD partition2: file3 file4 (day=1)
RDD partition3: file5 (day=2)
RDD partition4: file6 (dat-3)

**Comment by ulysses-you** [comment]
I think it can be merge partition one by one in a loop.
